### PR TITLE
Fixed 500 error when requesting include that didn't exist

### PR DESCRIPTION
**Pull request by cclep**
- Error was caused by the eager loading functionality
- Fix was to check for available includes if possible before blindly using them as an eager load
- This fix will quietly ignore invalid includes. We should consider throwing a http exception as a better fix

### CODE REVIEW

**Comment by theHarvester** [comment]
Bump. @hskrasek can this get reviewed please? Thanks mate

**Comment by hskrasek** [comment]
Is there an issue or something that this references? Not sure I fully understand the issue thats being fixed here

**Comment by cclep** [comment]
a 500 error was thrown if an include not present in the 'available includes' is requested. This fixes this error by checking first if the include is valid.

**Comment by hskrasek** [comment]
@cclep Do you have an example of that 500 error?

**Comment by cclep** [comment]
@hskrasek 
{
  "message": "Call to undefined method Illuminate\\Database\\Query\\Builder::doesntexist()",
  "statusCode": 500,
  "debug": {
    "line": 2101,
    "file": "/home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php",
    "class": "BadMethodCallException",
    "trace": [
      "#0 [internal function]: Illuminate\\Database\\Query\\Builder->__call('doesntexist', Array)",
      "#1 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(1015): call_user_func_array(Array, Array)",
      "#2 [internal function]: Illuminate\\Database\\Eloquent\\Builder->__call('doesntexist', Array)",
      "#3 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(3444): call_user_func_array(Array, Array)",
      "#4 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(545): Illuminate\\Database\\Eloquent\\Model->__call('doesntexist', Array)",
      "#5 [internal function]: Illuminate\\Database\\Eloquent\\Builder->Illuminate\\Database\\Eloquent\\{closure}()",
      "#6 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/Relation.php(171): call_user_func(Object(Closure))",
      "#7 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(546): Illuminate\\Database\\Eloquent\\Relations\\Relation::noConstraints(Object(Closure))",
      "#8 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(517): Illuminate\\Database\\Eloquent\\Builder->getRelation('doesntexist')",
      "#9 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(497): Illuminate\\Database\\Eloquent\\Builder->loadRelation(Array, 'doesntexist', Object(Closure))",
      "#10 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Collection.php(43): Illuminate\\Database\\Eloquent\\Builder->eagerLoadRelations(Array)",
      "#11 [internal function]: Illuminate\\Database\\Eloquent\\Collection->load(Array)",
      "#12 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Pagination/AbstractPaginator.php(475): call_user_func_array(Array, Array)",
      "#13 /home/vagrant/platform3/backend/vendor/dingo/api/src/Transformer/Adapter/Fractal.php(98): Illuminate\\Pagination\\AbstractPaginator->__call('load', Array)",
      "#14 /home/vagrant/platform3/backend/vendor/dingo/api/src/Transformer/Factory.php(87): Dingo\\Api\\Transformer\\Adapter\\Fractal->transform(Object(Illuminate\\Pagination\\LengthAwarePaginator), Object(Platform\\API\\V1\\Transformers\\TargetedCampaign\\ChangeAgentTransformer), Object(Dingo\\Api\\Transformer\\Binding), Object(Dingo\\Api\\Http\\Request))",
      "#15 /home/vagrant/platform3/backend/vendor/dingo/api/src/Http/Response.php(119): Dingo\\Api\\Transformer\\Factory->transform(Object(Illuminate\\Pagination\\LengthAwarePaginator))",
      "#16 /home/vagrant/platform3/backend/vendor/dingo/api/src/Routing/Router.php(603): Dingo\\Api\\Http\\Response->morph('json')",
      "#17 /home/vagrant/platform3/backend/vendor/dingo/api/src/Routing/Router.php(572): Dingo\\Api\\Routing\\Router->prepareResponse(Object(Dingo\\Api\\Http\\Response), Object(Dingo\\Api\\Http\\Request), 'json')",
      "#18 /home/vagrant/platform3/backend/vendor/dingo/api/src/Http/Middleware/Request.php(112): Dingo\\Api\\Routing\\Router->dispatch(Object(Dingo\\Api\\Http\\Request))",
      "#19 [internal function]: Dingo\\Api\\Http\\Middleware\\Request->Dingo\\Api\\Http\\Middleware\\{closure}(Object(Dingo\\Api\\Http\\Request))",
      "#20 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(139): call_user_func(Object(Closure), Object(Dingo\\Api\\Http\\Request))",
      "#21 /home/vagrant/platform3/backend/vendor/barryvdh/laravel-cors/src/HandlePreflight.php(25): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Dingo\\Api\\Http\\Request))",
      "#22 [internal function]: Barryvdh\\Cors\\HandlePreflight->handle(Object(Dingo\\Api\\Http\\Request), Object(Closure))",
      "#23 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(124): call_user_func_array(Array, Array)",
      "#24 /home/vagrant/platform3/backend/vendor/barryvdh/laravel-cors/src/HandleCors.php(34): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Dingo\\Api\\Http\\Request))",
      "#25 [internal function]: Barryvdh\\Cors\\HandleCors->handle(Object(Dingo\\Api\\Http\\Request), Object(Closure))",
      "#26 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(124): call_user_func_array(Array, Array)",
      "#27 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php(44): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Dingo\\Api\\Http\\Request))",
      "#28 [internal function]: Illuminate\\Foundation\\Http\\Middleware\\CheckForMaintenanceMode->handle(Object(Dingo\\Api\\Http\\Request), Object(Closure))",
      "#29 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(124): call_user_func_array(Array, Array)",
      "#30 [internal function]: Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Dingo\\Api\\Http\\Request))",
      "#31 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(102): call_user_func(Object(Closure), Object(Dingo\\Api\\Http\\Request))",
      "#32 /home/vagrant/platform3/backend/vendor/dingo/api/src/Http/Middleware/Request.php(113): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))",
      "#33 /home/vagrant/platform3/backend/vendor/dingo/api/src/Http/Middleware/Request.php(89): Dingo\\Api\\Http\\Middleware\\Request->sendRequestThroughRouter(Object(Dingo\\Api\\Http\\Request))",
      "#34 [internal function]: Dingo\\Api\\Http\\Middleware\\Request->handle(Object(Dingo\\Api\\Http\\Request), Object(Closure))",
      "#35 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(124): call_user_func_array(Array, Array)",
      "#36 [internal function]: Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))",
      "#37 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(102): call_user_func(Object(Closure), Object(Illuminate\\Http\\Request))",
      "#38 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(122): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))",
      "#39 /home/vagrant/platform3/backend/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(87): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter(Object(Illuminate\\Http\\Request))",
      "#40 /home/vagrant/platform3/backend/public/index.php(53): Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))",
      "#41 {main}"
    ]
  }
}

File: src/Transformer/Adapter/Fractal.php [hunk 1]
```diff
@@ -12,6 +12,7 @@
 use League\Fractal\Resource\Collection as FractalCollection;
 use Illuminate\Database\Eloquent\Collection as EloquentCollection;
 use Illuminate\Contracts\Pagination\Paginator as IlluminatePaginator;
+use League\Fractal\TransformerAbstract;
```
Comment (lucasmichot) [inline src/Transformer/Adapter/Fractal.php]: Import are ordered by length

File: src/Transformer/Adapter/Fractal.php [hunk 2]
```diff
@@ -65,7 +66,7 @@ public function __construct(FractalManager $fractal, $includeKey = 'include', $i
      * Transform a response with a transformer.
      *
      * @param mixed                          $response
-     * @param object                         $transformer
+     * @param object|TransformerAbstract     $transformer
```
Comment (lucasmichot) [inline src/Transformer/Adapter/Fractal.php]: Should be `@param League\Fractal\TransformerAbstract|object $transformer`

**Comment by georaldc** [comment]
One solution is to disable eager loading in Dingo and just handle it yourself. Not sure how viable that is though. Probably depends on your codebase

**Comment by theHarvester** [comment]
@georaldc that removes half of the benefit of using dingo instead of fractal directly...

**Comment by m1r0** [comment]
Got the same problem when using model bindings.

**Comment by m1r0** [comment]
Solved the problem by registering a custom exception:

```

**Comment by fbingha** [comment]
Just ran into this issue, the fix provided over 2 years ago resolves the issue for me.

**Comment by specialtactics** [comment]
PR Merged.

**Pull request merged**